docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,156 @@
+sketch of a generic and simple `groupService` implementation, to handle a
+desecrate stage or sequence of services.  As some services have
+dependencies during construction or initialization and as a starting
+point, a cluster of dependencies can be wrapped as a single
+`service.Service`.
+
+## Goals
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects. 
+  
+  - Reactors should not need to expose their interfaces *within* the
+    implementation of the node type, except in the context of some groups of
+    service. 
+
+- Expose some "unsafe"  way of replacing (some?) existing components, and or
+  inserting arbitrary reactors or services into the tendermint node's
+  initialization.
+  
+  - This may include creating an exported, top-level "unsafe" package with
+    several interfaces and constructors, or simply a function in the
+    node package. 
+
+- Refactor and simplify the process start/initialization logic,
+  separately from--but as a precursor to--permitting user replacement
+  of components.
+
+## Non-Goals
+
+- Fully abstract dependency-based process initialization. 
+
+- Fully abstract replacement of core tendermint components without vendoring.
+
+- Parallelized initialization of components.
+
+## Questions
+
+- To what extent does this new initialization framework need to accommodate
+  the legacy p2p stack? Would it be possible to delay a great deal of this
+  work to the 0.36 cycle to avoid this complexity? 
+  
+  - Answer: _depends on timing_, and the requirement to ship pluggable
+    reactors in 0.35. 
+  
+- There's clear and common interest in injecting some kind of mempool, and
+  intuitively this makes sense as the highest priority.
+  
+  - Does the existing `mempool.Mempool` interface make sense to expose and
+    commit to, or should we make a smaller interface? 
+    
+	- Answer: _yes_, smaller mempool interface is better. 
+
+  - Is it reasonable for users to bring their own mempool without bringing
+    their own reactor?
+	
+	- Answer: _probably not_, but this might be a function of how the
+      current implementations work and can be refactored. 
+
+- Are there reactors, other than the mempool, that make sense to support
+  ad-hoc replacing (e.g. statesync, blockchain (fastsync),) and if so should
+  we provide any first-class support for this?
+
+  - Answer: _defer_. Certainly this shouldn't be attempted until after
+    there is a refactor of the consensus reactor. Is more likely than
+    block/fast sync. 
+
+- There's a dependency (and nearly a cycle), between consensus,
+  blockchain (fastsync), which makes it very hard to allow meaningful
+  injection of any of these components without allowing injection of
+  all of them (statesync provider as a possible exception.) Is this
+  acceptable?
+  
+  - Answer: _yes_. 
```

Comment:
> and therefore it's not just startup order (as you model, I think correctly,) but also import/initialization order related.

right but if you fix the start up phases then none of these reactors need to have any dependency/reference to another; it's all managed by the node.

>  and nodes can move between "doing fastsync" and "doing consensus" if the node starts lagging.

This is a bit of a special case. The consensus reactor needs to signal to the node somehow that it's falling behind so that the node can stop it and briefly start up fast sync. As a caveat, this needs to only happen if the node operator has set fast sync to true. AFACIS, Either:
- The node needs to gain information on the height that it is at and the height that it's peers are at and from this information make the decision on whether to switch.
- The node needs to listen on the consensus reactor's `Quit()` channel (see `service.Service`) for when the consensus "gives up" and the node must switch to fast sync to catch up

Reply:
I think because this dependency related to consensus and bootstrapping (fast sync or otherwise) is not a footnote/edge case but actually the common case.

Reply:
Well currently it's not possible to fallback into fast sync so there's no case for it at the moment. I called it a special case because unlike the other start up processes where the node just runs it and awaits for it to be complete, here the node must monitor the state of the consensus reactor and make a decision. Hence, whilst the others may be more straight forward, adding this feature requires more sophisticated orchestration.

Reply:
Chiming in to say that I strongly agree that the node or some external control structure should handle the orchestration between fastsync and consensus. Personally I'd prefer to have the orchestrator be aware of height and kill consensus when it's too far behind to keep the responsibility of consensus as simple as possible (when running, try to produce block)